### ISSUE TITLE
Create parlai options to 1) create optimizer states on CPU rather than GPU and 2) disable saving optimizer states at checkpoints

### ISSUE THREAD

**Issue by sjscotti** [facebookresearch/ParlAI#4328]
**Is your feature request related to a problem? Please describe.**

I have been trying to train the larger BB2_3B model on my single CPU/single GPU desktop and been running into GPU memory limits and (when I can get it to fit in memory limits) is very slow.  I have a lot of CPU memory but only 8GB of GPU memory.    Also, for large models, the saved checkpoint is extremely large (>37GB for BB2_3B using an optimizer that saves 3 optimizer states) and the save operation is very slow due to saving the optimizer state.

**Describe the solution you'd like**
Modify the optimizers (e.g., in `fp16.py`) to allow an option to create the optimizer states on CPU rather than GPU, and read them over to GPU only as needed for calculations.  I have hard coded this for an optimizer I installed in parlai as described in the issues https://github.com/XuezheMax/apollo/issues/10#issuecomment-1022882758  .   Using this approach for optimizers in parlai through a command option would be helpful for users with limited hardware.

Modify the state saving step (e.g., `def state_dict(self):`  in `torch_agent.py`) with the option  to skip saving optimizer state dict.  

**Describe alternatives you've considered**
I've hard coded these suggestions myself and they seem to work well.   I need much less GPU memory and it runs optimizer steps faster (see issue comment mentioned earlier). However, it would be preferable if they can be options for the `parlai` command. These options may help others with limited hardware for using parlai

**Comment by stephenroller** [comment]
We could steal the --cpu-offload feature from fairseq/fairscale.

**Comment by sjscotti** [comment]
I think the --cpu-offload feature is a good thing to implement.  I also found that it was just about effective computationally to remove the optimizer states from a checkpoint and restart with a 50 step warmup as starting from the checkpointed state when the models are big and you save the state often.  So if there were a way to add an option to not save optimizer state, it would be helpful too.
